### ISSUE TITLE
SIGSEGV on unaligned `vmovdqa` if compiled with AVX-512

### ISSUE THREAD

**Issue by Gumix** [tarantool/tarantool#10215]
**Steps to reproduce**

<details>
  <summary>Tarantool version</summary>

  ```
Tarantool 3.2.0-entrypoint-116-gbf01fb200
Target: Linux-x86_64-RelWithDebInfo
Build options: cmake . -DCMAKE_INSTALL_PREFIX=/usr/local -DENABLE_BACKTRACE=TRUE
Compiler: GNU-13.3.0
C_FLAGS:-march=skylake-avx512 -fexceptions -funwind-tables -fasynchronous-unwind-tables -fno-common -msse2 -Wformat -Wformat-security -Werror=format-security -fstack-protector-strong -fPIC -std=c11 -Wall -Wextra -Wno-gnu-alignof-expression -fno-gnu89-inline -Wno-cast-function-type -O2 -g -DNDEBUG -ggdb -O2
CXX_FLAGS:-march=skylake-avx512 -fexceptions -funwind-tables -fasynchronous-unwind-tables -fno-common -msse2 -Wformat -Wformat-security -Werror=format-security -fstack-protector-strong -fPIC -std=c++11 -Wall -Wextra -Wno-invalid-offsetof -Wno-gnu-alignof-expression -Wno-cast-function-type -O2 -g -DNDEBUG -ggdb -O2
  ```
</details>

<details>
  <summary>GCC version</summary>

  ```
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/libexec/gcc/x86_64-linux-gnu/13/lto-wrapper
OFFLOAD_TARGET_NAMES=nvptx-none:amdgcn-amdhsa
OFFLOAD_TARGET_DEFAULT=1
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Debian 13.3.0-1' --with-bugurl=file:///usr/share/doc/gcc-13/README.Bugs --enable-languages=c,ada,c++,go,d,fortran,objc,obj-c++,m2 --prefix=/usr --with-gcc-major-version-only --program-suffix=-13 --program-prefix=x86_64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/libexec --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --enable-bootstrap --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-libstdcxx-backtrace --enable-gnu-unique-object --disable-vtable-verify --enable-plugin --enable-default-pie --with-system-zlib --enable-libphobos-checking=release --with-target-system-zlib=auto --enable-objc-gc=auto --enable-multiarch --disable-werror --enable-cet --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-offload-targets=nvptx-none=/build/reproducible-path/gcc-13-13.3.0/debian/tmp-nvptx/usr,amdgcn-amdhsa=/build/reproducible-path/gcc-13-13.3.0/debian/tmp-gcn/usr --enable-offload-defaulted --without-cuda-driver --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu --with-build-config=bootstrap-lto-lean --enable-link-serialization=3
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 13.3.0 (Debian 13.3.0-1)
  ```
</details>

<details>
  <summary>How to build</summary>

  ```bash
$ mkdir build
$ cd build
$ cmake .. -DCMAKE_BUILD_TYPE=RelWithDebInfo -DCMAKE_C_FLAGS=-march=skylake-avx512 -DCMAKE_CXX_FLAGS=-march=skylake-avx512
$ make
  ```
</details>

How to run:

```bash
$ tarantool -e "box.cfg{}"
2024-07-09 02:02:02.935 [0] main/104/interactive I> Tarantool 3.2.0-entrypoint-116-gbf01fb200 Linux-x86_64-RelWithDebInfo
2024-07-09 02:02:02.935 [0] main/104/interactive I> log level 5 (INFO)
2024-07-09 02:02:02.935 [0] main/104/interactive I> wal/engine cleanup is paused
2024-07-09 02:02:02.935 [0] main/104/interactive I> mapping 268435456 bytes for memtx tuple arena...
2024-07-09 02:02:02.935 [0] main/104/interactive I> Actual slab_alloc_factor calculated on the basis of desired slab_alloc_factor = 1.044274
2024-07-09 02:02:02.935 [0] main/104/interactive I> mapping 134217728 bytes for vinyl tuple arena...
Segmentation fault
  code: 128
  addr: (nil)
[...]
#1  0x563a6c9837d0 in crash_signal_cb+160
#2  0x7fd0d1e58580 in __sigaction+64
#3  0x563a6c7cfe11 in iproto_init+785
#4  0x563a6c89e82d in box_cfg+637
#5  0x563a6c92d029 in lbox_cfg_load(lua_State*)+9
#6  0x563a6c9c4ca3 in lj_BC_FUNCC+70
#7  0x563a6c9cb882 in lua_pcall+114
#8  0x563a6c96835b in luaT_call+11
#9  0x563a6c95d0fd in run_script_f+1309
#10 0x563a6c7c41bc in fiber_cxx_invoke(int (*)(__va_list_tag*), __va_list_tag*)+12
#11 0x563a6c98d7a6 in fiber_loop+70
#12 0x563a6cbfa058 in coro_init+72
```

**Actual behavior**

Tarantool will crash or not, depending on the alignment of `iproto_threads`, allocated by:

https://github.com/tarantool/tarantool/blob/bf01fb200777c825e54b451fe3e85d3de4d85532/src/box/iproto.cc#L3867-L3868

> [!NOTE]
> On x86-64 Linux the 16-byte alignment is guaranteed, however the pointer may be accidentally 32-byte aligned.

Tarantool crashes here:

https://github.com/tarantool/tarantool/blob/bf01fb200777c825e54b451fe3e85d3de4d85532/src/box/iproto.cc#L3640-L3641

on the instruction that performs a store to the 256-bit (32-byte) aligned address:

```asm
vmovdqa YMMWORD PTR [r14 + 0xb40], ymm0
```

where `r14` points to `iproto_thread` and `dml_route[1]` has offset 0xb40 (2880), which is multiple of 32.

The compiler vectorized this access using 256-bit vector instructions. Here is GCC RTL dump from `iproto.cc.cc.339r.final`, note `A256`:

```lisp
(insn:TI 238 683 240 6 (set (mem:V4DI (plus:DI (reg/v/f:DI 43 r15 [orig:108 iproto_thread ] [108])
(const_int 2880 [0xb40]))
[62 MEM <vector(4) long unsigned int> [(struct cmsg_hop * *)iproto_thread_70 + 2880B]+0 S32 A256])
(reg:V4DI 20 xmm0 [184])) "./src/box/iproto.cc":3641:30 1835 {movv4di_internal}
(expr_list:REG_DEAD (reg:V4DI 20 xmm0 [184])
(nil)))
```

But why does the compiler assume that the base address is 256-bit aligned? The problem comes from the nested unnamed struct that causes over-alignment of the parent `iproto_thread` struct.

https://github.com/tarantool/tarantool/blob/bf01fb200777c825e54b451fe3e85d3de4d85532/src/box/iproto.cc#L245-L251

From the `iproto.cc.cc.001l.class` dump:

```
Class iproto_thread::<unnamed struct>
   size=64 align=64
   base size=64 base align=64
iproto_thread::<unnamed struct> (0x0x7fc361f64000) 0

Class iproto_thread
   size=3840 align=64
   base size=3840 base align=64
iproto_thread (0x0x7fc361f13f60) 0
```

If I remove `alignas(CACHELINE_SIZE)`, the `iproto_thread` struct is no longer considered as aligned, so there are no crashes, because in that case `vmovdqu` is generated instead of `vmovdqa`.

```
Class iproto_thread::<unnamed struct>
   size=16 align=8
   base size=16 base align=8
iproto_thread::<unnamed struct> (0x0x7f1688214000) 0

Class iproto_thread
   size=3776 align=8
   base size=3776 base align=8
iproto_thread (0x0x7f16881c3f60) 0
```

When `calloc()` is replaced by `new[]`, there is a warning indicating that something goes wrong:

```
./src/box/iproto.cc: In function ‘void iproto_init(int)’:
./src/box/iproto.cc:3869:57: warning: ‘new’ of type ‘iproto_thread’ with extended alignment 64 [-Waligned-new=]
 3869 |         iproto_threads = new iproto_thread[threads_count];
      |                                                         ^
./src/box/iproto.cc:3869:57: note: uses ‘void* operator new [](long unsigned int)’, which does not have an alignment parameter
./src/box/iproto.cc:3869:57: note: use ‘-faligned-new’ to enable C++17 over-aligned new support
```

**TL;DR**

> [!CAUTION]
> Do not allocate C++ structs with `malloc()`.
